File: C++ Codes/LinkedList/LinkedList.cpp
// LinkedList.cpp : This file contains the 'main' function. Program execution begins and ends there.
//

#include <iostream>
using namespace std;
struct Node
{
    int data;
    Node* next;
};

class linkedList
{
private:
    Node* head, *tail;
public:
    linkedList()
    {
        head = NULL;
        tail = NULL;
    }
    void addNode(int n)
    {
        Node* temp = new Node;
        temp->data = n;
        temp->next = NULL;

        if (head == NULL)
        {
            head = temp;
            tail = temp;
        }
        else
        {
            tail->next = temp;
            tail = temp;
        }
    }

    void addAtStart(int val)
    {
        Node* temp = new Node;
        temp->data = val;
        temp->next = head;
        head = temp;
    }

    void printList()
    {
        Node* currNode = head;

        while (currNode)
        {
            cout << "\n" << currNode->data;
            currNode = currNode->next;
        }
    }

    void addAtEnd(int val)
    {
        Node* temp = new Node;
        temp->data = val;
        temp->next = NULL;
        tail->next = temp;
        tail = temp;
    }

    void addAtPosition(int pos, int val)
    {
        Node* temp = new Node();
        Node* cur = new Node();
        Node* prev = new Node();

        cur = head;

        for (int i = 1; i < pos; i++)
        {
            prev = cur;
            cur = cur->next;
        }

        temp->data = val;
        prev->next = temp;
        temp->next = cur;
    }

    void deleteAtStart()
    {
        Node* temp = new Node();
        temp = head;
        head = head->next;
        delete temp;
    }

    void deleteAtEnd()
    {
        Node* cur = new Node();
        Node* prev = new Node();
        cur = head;
        while (cur->next != NULL)
        {
            prev = cur;
            cur = cur->next;
        }
        prev->next = NULL;
        delete cur;
        tail = prev;

    }

    void deleteAtPosition(int pos)
    {
        Node* prev = new Node();
        Node* cur = new Node();

        cur = head;
        for (int i = 1; i < pos; i++)
        {
            prev = cur;
            cur = cur->next;
        }
        prev->next = cur->next;
    }

};
int main()
{
    std::cout << "Hello World!\n";
    linkedList lst;
    lst.addNode(123);
    lst.addNode(23);
    lst.addNode(75);
    lst.addAtStart(01);
    lst.addAtEnd(999);
    lst.addAtPosition(3, 50);
    //lst.deleteAtStart();
    //lst.deleteAtEnd();
    lst.deleteAtPosition(3);
    lst.printList();
}

File: javascript tutorials/LeetCode/linkedList.js
function Node(val)
{
  this.val = val;
  this.next = null;
}

function LinkedList()
{
  this.count = 0;
  this.head = null;
  this.tail = null;


  this.addToStart = function(val)
  {
    if(this.head == null && this.tail == null)
    {
      var curr = new Node(val);
      this.head = curr;
      this.tail = curr;
    }
    else
    {
      var curr = new Node(val);
      curr.next = this.head;
      this.head = curr;
    }
    this.count++;
  }

  this.addToEnd = addToEnd;
  this.remove = remove;
}

function addToEnd(val)
{
  if(this.head == null && this.tail == null)
  {
    var curr = new Node(val);
    this.head = curr;
    this.tail = curr;
  }
  else
  {
    var curr = new Node(val);
    this.tail.next = curr;
    this.tail = curr;
  }
  this.count++;
}

function remove(val)
{
  var c = this.head;
  var p = null;
  while(c!= null)
  {
    if(p == null)
    {
      if(c.val == val)
      {
        l = c.next;
        c = l;
      }
      else
      {
        p = c;
        c = c.next;
      }
    }
    else
    {
      if(c.val == val)
      {
        p.next = c.next;
        c = c.next;
      }
      else
      {
        p = c;
        c = c.next;
      }
    }
  }
}


var l = new LinkedList();
l.addToStart(5);
l.addToStart(6);
l.addToStart(7);
l.addToEnd(3);
l.addToEnd(10);
l.remove(10);
var node = l.head;
while(node.next != null)
{
  console.log(node.val);
  node = node.next;
}

if(node.next == null)
console.log(node.val);